File: JAVA-DSA/BasicMath.java
import java.util.Vector;
import java.util.Collections;

class BasicMath {
    public static int countDigits(int num) {
        if ( num == 0 ) return 1;
        num = Math.abs(num);
        int count = 0;
        while ( num > 0 ) {
            int lastDigit = num % 10;
            count = count + 1;
            num = num / 10;
        }
        return count;
    }

    public static int reverseNumber(int num) {
        int revNum = 0;
        while ( num > 0 ) {
            int lastDigit = num % 10;
            revNum = (revNum * 10) + lastDigit;
            num = num / 10;
        }
        return revNum;
    }

    public static boolean palindrome(int num) {
        if (num < 0 ) return false;
        int originalNum = num;
        int revNum = 0;
        while ( num > 0 ) {
            int lastDigit = num % 10;
            revNum = (revNum * 10) + lastDigit;
            num = num / 10;
        }
        return originalNum == revNum;
    }

    public static boolean armstrongNumber(int num) {
        if (num < 0 ) return false;
        int sum = 0;
        int originalNum = num;
        int numberOfDigits = String.valueOf(num).length();
        while ( num > 0 ) {
            int lastDigit = num % 10;
            sum += Math.pow(lastDigit, numberOfDigits);
            num = num / 10;
        }
        return sum == originalNum;
    }

    public static void printDivisors(int num) {
        for ( int i = 1; i <= num; i++ ){
            if ( num%i == 0 ) {
                System.out.print(i + " ");
            }
        }
    }

    public static void printDivisorsOther(int num) {
        Vector<Integer> ls = new Vector<Integer>();
        for ( int i = 1; i*i <= num; i++ ){
            if ( num%i == 0 ) {
                ls.add(i);
                if ( (num/i) != i ) {
                    ls.add(num/i);
                }
            }
        }
        Collections.sort(ls);
        System.out.println(ls);
    }

    public static void primeNumber(int num) {
        int count = 0;
        for ( int i = 1; i*i <= num; i++ ){
            if ( num%i == 0 ) {
                count++;
                if ( (num/i) != i ) {
                    count++;
                }
            }
        }
        if ( count == 2 ) System.out.println("prime num : " + num + " true");
        else System.out.println("prime num: " + num + " false");
    }

    public static void main(String args[]) {
        int num = 7789;
        int num2 = 141;
        int num3 = 1634;
        System.out.println("count digits -> " + countDigits(num) );
        System.out.println("reverse number -> " + reverseNumber(num));
        System.out.println("palindrome : " + num2 + " -> " + palindrome(num2));
        System.out.println("Armstrong number : " + num3 + " -> " + armstrongNumber(num3));
        System.out.println("Divisors of number" + num2);
        printDivisors(num2);
        System.out.println();
        printDivisorsOther(num2);
        primeNumber(17);
    }
}

File: DSA-CPP/BasicMath.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>
using namespace std;

class BasicMath {
  public:
    static int countDigits(int num) {
      if (num == 0) return 1;
      num = abs(num);
      int count = 0;
      while (num > 0) {
        num = num / 10;
        count++;
      }
      return count;
    }

    static int reverseNumber(int num) {
      int revNum = 0;
      while (num > 0) {
        int lastDigit = num % 10;
        revNum = (revNum * 10) + lastDigit;
        num = num / 10;
      }
      return revNum;
    }

    static bool palindrome(int num) {
      if (num < 0) return false;
      int originalNum = num;
      int revNum = 0;
      while (num > 0) {
        int lastDigit = num % 10;
        revNum = (revNum * 10) + lastDigit;
        num = num / 10;
      }
      return originalNum == revNum;
    }

    static bool armstrongNumber(int num) {
      if (num < 0) return false;
      int sum = 0;
      int originalNum = num;
      int numberOfDigits = to_string(num).length();
      while (num > 0) {
        int lastDigit = num % 10;
        sum += pow(lastDigit, numberOfDigits);
        num = num / 10;
      }
      return sum == originalNum;
    }

    static void printDivisors(int num) {
      for (int i = 1; i <= num; i++) {
        if (num % i == 0) {
          cout << i << " ";
        }
      }
      cout << endl;
    }

    static void printDivisorsOther(int num) {
      vector<int> ls;
      for (int i = 1; i * i <= num; i++) {
        if (num % i == 0) {
          ls.push_back(i);
          if ((num / i) != i) {
            ls.push_back(num / i);
          }
        }
      }
      sort(ls.begin(), ls.end());
      for (int i = 0; i < ls.size(); i++) {
        cout << ls[i] << " ";
      }
      cout << endl;
    }

    static void primeNumber(int num) {
      int count = 0;
      for (int i = 1; i * i <= num; i++) {
        if (num % i == 0) {
          count++;
          if ((num / i) != i) {
            count++;
          }
        }
      }
      if (count == 2) cout << "prime num : " << num << " true" << endl;
      else cout << "prime num : " << num << " false" << endl;
    }
};

int main() {
  int num = 7789;
  int num2 = 141;
  int num3 = 1634;
  cout << "count digits -> " << BasicMath::countDigits(num) << endl;
  cout << "reverse number -> " << BasicMath::reverseNumber(num) << endl;
  cout << "palindrome : " << num2 << " -> " << BasicMath::palindrome(num2) << endl;
  cout << "Armstrong number : " << num3 << " -> " << BasicMath::armstrongNumber(num3) << endl;
  cout << "Divisors of number " << num2 << ": ";
  BasicMath::printDivisors(num2);
  cout << "Other divisors of number " << num2 << ": ";
  BasicMath::printDivisorsOther(num2);
  BasicMath::primeNumber(17);
  return 0;
}

